src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIResponseChatClient.cs

```diff
@@ -0,0 +1,582 @@
+            }
+            else if (options.ResponseFormat is ChatResponseFormatJson jsonFormat)
+            {
+                result.TextOptions.ResponseFormat = jsonFormat.Schema is { } jsonSchema ?
+                    ResponseTextFormat.CreateJsonSchemaFormat(
+                        jsonFormat.SchemaName ?? "json_schema",
+                        BinaryData.FromBytes(JsonSerializer.SerializeToUtf8Bytes(jsonSchema, OpenAIJsonContext.Default.JsonElement)),
+                        jsonFormat.SchemaDescription) :
+                    ResponseTextFormat.CreateJsonObjectFormat();
+            }
+        }
+
+        return result;
+    }
+
+    /// <summary>Convert a sequence of <see cref="ChatMessage"/>s to <see cref="ResponseItem"/>s.</summary>
+    private static IEnumerable<ResponseItem> ToOpenAIResponseItems(
+        IEnumerable<ChatMessage> inputs, JsonSerializerOptions options)
+    {
+        foreach (ChatMessage input in inputs)
+        {
+            if (input.Role == ChatRole.System ||
+                input.Role == _chatRoleDeveloper)
+            {
+                string text = input.Text;
+                if (!string.IsNullOrWhiteSpace(text))
+                {
+                    yield return input.Role == ChatRole.System ?
+                        ResponseItem.CreateSystemMessageItem(text) :
+                        ResponseItem.CreateDeveloperMessageItem(text);
+                }
+
+                continue;
+            }
+
+            if (input.Role == ChatRole.User)
+            {
+                yield return ResponseItem.CreateUserMessageItem(ToOpenAIResponsesContent(input.Contents));
+                continue;
+            }
+
+            if (input.Role == ChatRole.Tool)
+            {
+                foreach (AIContent item in input.Contents)
+                {
+                    switch (item)
+                    {
+                        case FunctionResultContent resultContent:
+                            string? result = resultContent.Result as string;
+                            if (result is null && resultContent.Result is not null)
+                            {
+                                try
+                                {
+                                    result = JsonSerializer.Serialize(resultContent.Result, options.GetTypeInfo(typeof(object)));
+                                }
+                                catch (NotSupportedException)
+                                {
+                                    // If the type can't be serialized, skip it.
+                                }
+                            }
+
+                            yield return ResponseItem.CreateFunctionCallOutputItem(resultContent.CallId, result ?? string.Empty);
+                            break;
+                    }
+                }
+
+                continue;
+            }
+
+            if (input.Role == ChatRole.Assistant)
```

Comment:
What's the situation where the outbound data includes "assistant" messages, given the server statefulness?

I think this probably relates to the fact that even with server statefulness, you can freely prepopulate it with any conversation history you like. Is that right?

If so I see that we're only supporting `TextContent` and `FunctionCallContent` in an outbound "Assistant" message. Longer term should we expect to generalize this so that, for example, you can preseed the conversation with some `DataContent` that an assistant might previously have returned? Or perhaps more likely, citations, URLs, etc.

Reply:
> What's the situation where the outbound data includes "assistant" messages, given the server statefulness?

You don't have to use the server history. If you don't specify a previous response id, it behaves just like chat completion. And if you do, it's so just concatenating the provided messages into whatever it already has (which, if there's no previous ID, it's nothing). So you could have part of a conversion stored on the server and send another chunk of it.

> If so I see that we're only supporting TextContent and FunctionCallContent in an outbound "Assistant" message.

That's all the underlying API supports right now.